### PR TITLE
Xbox NXDK platform

### PR DESCRIPTION
**Pull request by glebm**
DevilutionX now builds with https://github.com/XboxDev/nxdk.

xbe_logo

Reported 26 FPS on a real Xbox in town.

There are some outstanding PRs to fix a few issues (listed below), so we build from my fork for now.

Note: Requires clang that comes with Ubuntu 22.04 (20.04 won't work).

Issues:

* [x] https://github.com/XboxDev/nxdk-libcxx/issues/9 and https://github.com/XboxDev/nxdk/issues/587
  * Fix for pdclib by NXDK folks: https://github.com/XboxDev/nxdk-pdclib/pull/52
  * Fix for nxdk by NXDK folks: https://github.com/XboxDev/nxdk/pull/603
  * Fix for libcxx: https://github.com/XboxDev/nxdk-libcxx/pull/8
* [x] https://github.com/XboxDev/nxdk/issues/594
  * Port branch: https://github.com/glebm/fmt/tree/nxdk
  * Port PR: https://github.com/XboxDev/nxdk/pull/601
* [x] https://github.com/mackron/dr_libs/issues/237
* [x] https://github.com/realnc/SDL_audiolib/issues/36
* [x] https://github.com/XboxDev/nxdk/pull/583
* [x] https://github.com/XboxDev/nxdk/issues/600
  * Fixed by NXDK folks: https://github.com/XboxDev/nxdk-sdl/pull/43
* [x] https://github.com/XboxDev/nxdk/issues/598
  *  Fix merged: https://github.com/XboxDev/nxdk-pdclib/pull/51
* [x] https://github.com/XboxDev/nxdk/issues/595
  * Fix merged: https://github.com/XboxDev/nxdk/pull/596
* [x] Missing POSIX headers: NXDK isn't POSIX, fixed on our end.
* [x] CMake toolchain pkg-config support
  * Fix merged: https://github.com/XboxDev/nxdk/pull/586
  * Fix merged: https://github.com/XboxDev/nxdk/pull/590

### CODE REVIEW

**Comment by AJenbo** [comment]
Wow, it would be exciting to finally have an OG Xbox port under our wings :)

**Comment by AJenbo** [comment]
@brentdc-nz  you might be interested in this :)

**Comment by glebm** [comment]
@Halofreak1990 I've noticed your port of Devilution (not DevilutionX) to the OG Xbox at https://github.com/Halofreak1990/devilution-xbox. This is a port of DevilutionX to the OG Xbox using NXDK.

**Comment by Halofreak1990** [comment]
> @Halofreak1990 I've noticed your port of Devilution (not DevilutionX) to the OG Xbox at https://github.com/Halofreak1990/devilution-xbox. This is a port of DevilutionX to the OG Xbox using NXDK that you might be interested in.

Could be interesting indeed. The reason I went with the Devilution codebase for my port and tried emulating all the Windows stuff, was because DevilutionX depends on SDL2, among other things, and I do not have access to a port of that for the OG Xbox, and I'm not comfortable (like, not at all) with SDL to even try attempting swapping to the SDL 1.x port that does exist. There's also the bit with the modern C++ stuff that Visual Studio 2003--which, up until NXDK came around, was required to build OG Xbox games--just doesn't take.

And then I got stuck on trying to wrap my head around an implementation of the Win32 dialog API, though I did manage to include the resources in the executable and access them. I just didn't get around to rendering anything.

**Comment by glebm** [comment]
@Halofreak1990 Luckily NXDK comes with SDL2. I've had a few issues porting it as well but thankfully NXDK folks were very helpful on GitHub and their Discord. This port should hopefully work at this point, you can try it out here:
https://github.com/diasurgical/devilutionX/suites/7238219081/artifacts/290474121

File: Source/utils/display.cpp [hunk 1]
```diff
@@ -235,6 +239,18 @@ bool SpawnWindow(const char *lpWindowName)
...
 	scePowerSetArmClockFrequency(444);
 #endif
+#ifdef NXDK
+	{
+		Size windowSize = *sgOptions.Graphics.resolution;
+		VIDEO_MODE xmode;
+		void *p = nullptr;
+		while (XVideoListModes(&xmode, 0, 0, &p)) {
+			if (windowSize.width >= xmode.width && windowSize.height == xmode.height)
+				break;
```
Comment (AJenbo) [inline Source/utils/display.cpp:249]: This logic could be better, but lets wait until XNDK gets the SDL calls implemented.

**Review by AJenbo — APPROVED** [review]

**Comment by AJenbo** [comment]
@glebm will you create an issue to track the remaning issues?

**Comment by AJenbo** [comment]
Also I guess we should look at packaging, but it seams it would be pretty trivial?

**Comment by AJenbo** [comment]
looks like everything is handled upstream. should we update things?

**Comment by glebm** [comment]
Still need https://github.com/XboxDev/nxdk/pull/606, and then I guess we'll need to fork libfmt into diasurgical

> Also I guess we should look at packaging, but it seams it would be pretty trivial?

The package is just a zip file that we already produce

**Comment by glebm** [comment]
Switched to our libfmt fork in https://github.com/diasurgical/devilutionX/pull/5011

File: .github/workflows/xbox_nxdk.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  xbox:
+    # `ubuntu-latest` is currently 20.04 but we require a newer clang.
+    runs-on: ubuntu-22.04
+    env:
+      NXDK_DIR: /opt/nxdk
+    steps:
+    - name: Install APT packages
+      run: sudo apt-get update && sudo apt-get install -y clang llvm lld bison flex cmake git gettext
+
+    - name: Clone NXDK Repo
```
Comment (JayFoxRox) [inline .github/workflows/xbox_nxdk.yml:20]: nit: The preferred spelling of nxdk is "nxdk" (all lowercase, even as first word in sentence).
Initially there were some disagreements about that spelling, but we settled on the all-lowercase variant a couple of years ago (if I remember correctly, also to differentiate it from the MS XDK). All-caps `NXDK` is only used in code for defines and such.

There were other instances in this PR which use the all-caps spelling, too.

*Slightly related: While MS didn't agree on a stylization around 2001, the XboxDev community prefers to use the now-preferred MS spelling of "Xbox" (capital "X", lower-case "box") for the console name. If there might be confusion with other consoles, "original Xbox" is used ("original" in lowercase unless at start of sentence or similar).*

Comment (glebm) [reply .github/workflows/xbox_nxdk.yml:20]: Thanks, noted!

**Pull request merged**